src/core/mesh/qgsmeshlayerutils.h

```diff
@@ -67,6 +67,30 @@ class CORE_EXPORT QgsMeshLayerUtils
       int valueIndex,
       int count );
 
+    /**
+     * \brief Returns gridded vector values, if extentInMap is default, uses the triangular mesh extent
+     *
+     * When calling, the size parameter could contain the maximum grid size wanted, if invalid (==QSize()), there is no limit.
+     * After return, this reference contains the effective grid size.
+     *
+     * \param meshLayer pointer to the mesh layer
+     * \param index dataset index
+     * \param xSpacing the spacing on X coordinate in map unit
+     * \param ySpacing the spacing on Y coordinate in map unit
+     * \param size contains the size (count of rows and columns) of the grid supporting the vectors
+     * \param minCorner coordinates of the corner with x minimum and y miminum
+     * \param extentInMap the extent where the grid is built in map coordinates
+     * \returns vectors on a grid, empty if dataset is no vector values
+     *
+     * \since QGIS 3.14
+     */
+    static QVector<QgsVector> griddedVectorValues( const QgsMeshLayer *meshLayer,
+        const QgsMeshDatasetIndex index,
+        double xSpacing,
+        double ySpacing,
+        QSize &size, QgsPointXY &minCorner,
+        const QgsRectangle extentInMap = QgsRectangle() );
+
```

Comment:
for this whole function. I am wondering if it is not the same as function void QgsMeshVectorArrowRenderer::drawVectorDataOnGrid( const QList<int> &trianglesInExtent ) https://github.com/qgis/QGIS/blob/master/src/core/mesh/qgsmeshvectorrenderer.cpp#L302

and if the "calculation" part cannot be shared. or is the logic different?

Reply:
Good point. It deserves to be try. Especially that could be used for the export vectors ticket.

Reply:
Finally, logic is not the same. For vector arrow rendering, the entry data is all the triangles in the extent, retrieve with spatial index. Each triangles is then treated to render the vector that are in the triangle. Here, the order of drawing vector is not important and the drawing operation is done in the loops, they are not stored.
For 3D, as the gridded vectors will be used for a texture, we need them stored and sorted. Maybe to export gridded vector it is the same logic.